pennylane/ops/op_math/composite.py

```diff
@@ -0,0 +1,226 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This submodule defines a base class for composite operations.
+"""
+import abc
+
+import numpy as np
+
+import pennylane as qml
+from pennylane import math
+from pennylane.operation import Operator
+
+
+class CompositeOp(Operator, abc.ABC):
+    """A base class for operators that are composed of other operators.
+
+    Args:
+        operands: (tuple[~.operation.Operator]): a tuple of operators which will be combined.
+
+    Keyword Args:
+        do_queue (bool): determines if the operator will be queued. Default is True.
+        id (str or None): id for the operator. Default is None.
+
+    The child composite operator should define the `_name` and `_op_symbol` properties
+    during initialization and define any relevant representations, such as
+    :meth:`~.operation.Operator.matrix` and :meth:`~.operation.Operator.decomposition`.
+    """
+
+    _eigs = {}  # cache eigen vectors and values like in qml.Hermitian
+    _op_symbol = None
+    _name = None
```

Comment:
Maybe instead of raising errors in `__init__`, we can make these abstract properties:
```
    @property
    @abc.abstractmethod
    def _op_symbol(self):
        pass
```

We might also be able to leave `_name` as just `self.__class__.__name__`, as is set in `Operator`.

Reply:
makes sense to me! updated